source/types.ts

```diff
@@ -0,0 +1,274 @@
+
+	/**
+	 * Method to decrement a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 */
+	decrement: (key: string) => Promise<void> | void
+
+	/**
+	 * Method to reset a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 */
+	resetKey: (key: string) => Promise<void> | void
+
+	/**
+	 * Method to reset everyone's hit counter.
+	 */
+	resetAll?: () => Promise<void> | void
+}
+
+/**
+ * The configuration options for the rate limiter.
+ */
+export interface Options {
+	/**
+	 * How long we should remember the requests.
+	 */
+	readonly windowMs: number
+
+	/**
+	 * The maximum number of connection to allow during the `window` before
+	 * rate limiting the client.
+	 *
+	 * Can be the limit itself as a number or express middleware that parses
+	 * the request and then figures out the limit.
+	 */
+	readonly max: number | ValueDeterminingMiddleware<number>
+
+	/**
+	 * The response body to send back when a client is rate limited.
+	 */
+	readonly message: any
+
+	/**
+	 * The HTTP status code to send back when a client is rate limited.
+	 *
+	 * Defaults to `HTTP 429 Too Many Requests` (RFC 6585).
+	 */
+	readonly statusCode: number
+
+	/**
+	 * Whether to send `X-RateLimit-*` headers with the rate limit and the number
+	 * of requests.
+	 */
+	readonly legacyHeaders: boolean
+
+	/**
+	 * Whether to enable support for the rate limit standardization headers (`RateLimit-*`).
+	 */
+	readonly standardHeaders: boolean
+
+	/**
+	 * The name of the property on the request object to store the rate limit info.
+	 *
+	 * Defaults to `rateLimit`.
+	 */
+	readonly requestPropertyName: string
+
+	/**
+	 * If `true`, the library will (by default) skip all requests that have a 4XX
+	 * or 5XX status.
+	 */
+	readonly skipFailedRequests: boolean
+
+	/**
+	 * If `true`, the library will (by default) skip all requests that have a
+	 * status code less than 400.
+	 */
+	readonly skipSuccessfulRequests: boolean
+
+	/**
+	 * Method to determine whether or not the request counts as 'succesful'. Used
+	 * when either `skipSuccessfulRequests` or `skipFailedRequests` is set to true.
+	 */
+	readonly requestWasSuccessful: ValueDeterminingMiddleware<boolean>
+
+	/**
+	 * Method to generate custom identifiers for clients.
+	 *
+	 * By default, the client's IP address is used.
+	 */
+	readonly keyGenerator: ValueDeterminingMiddleware<string>
+
+	/**
+	 * Method (in the form of middleware) to determine whether or not this request
+	 * counts towards a client's quota.
+	 */
+	readonly skip: ValueDeterminingMiddleware<boolean>
+
+	/**
+	 * Express request handler that sends back a response when a client is
+	 * rate-limited.
+	 */
+	readonly handler: RateLimitEventCallback
+
+	/**
+	 * Express request handler that sends back a response when a client has
+	 * reached their rate limit, and will be rate limited on their next request.
+	 */
+	readonly onLimitReached: RateLimitEventCallback
```

Comment:
I think we'll have to split these into different interfaces to keep `next` in the handler callback. Sorry :/

Reply:
No problem :D, will do

Reply:
Should we also pass the `next` param to the `onLimitReached` callback? For example, the user might want to skip setting a particular header if the `verbose` query param is `false`.

Reply:
No, `onLimitReached` is kind of a weird function. 

If your limit is 3 and an end-user makes 10 requests, `handler` will be called 7 times, but `onLimitReached` will only be called once. 

It's intended for logging that a user hit the limit, without logging every single request after the limit was hit.

Reply:
> If your limit is 3 and an end-user makes 10 requests, handler will be called 7 times, but onLimitReached will only be called once.

Yes, I understand that.

The use case I was describing was the server returning a field warning the client that their rate limit has been reached in the response, e.g.:

```jsonc
// -- response body

{
	"meta": {
		"status": 200,
		"rateLimitReached": true,
	},
	"data": {
		// ...
	}
}
```

Reply:
OK, let me think about that a bit. I feel like something won't work, but I can't quite put my finger on it.

Reply:
I figured out what was throwing me off: It sounds like you're wanting to give extra info on the last request that's allowed through - the 3rd request if the limit is 3.  Is that right?

However, `onLimitReached` is triggered on the 4th request. So, it's too late for a warning, the user is already blocked at that point. They won't get the `data`, just whatever the express-rate-limit `handler` function returns.

I think your use-case would be better served by checking the `req.rateLimit` property after `express-rate-limit` has done it's thing. 

I think `onLimitReached` predates `req.rateLimit` and, honestly, it should probably just be deprecated. It's very confusing and not particularly helpful. Everything that can be done with `onLimitReached` can also be done with a custom `handler` and an `if` statement.

Or, were you saying the 4th request should actually be allowed through, just with a warning? In that case, you could achieve the same effect by doing it in `handler` but bumping the limit up to `4`.

(To be clear, I think we should keep it in there with the same behavior for now, but just mark it as deprecated try to discourage people from using it.) 

A better replacement might be to return an EventEmitter and have some kind of event for this, but I'm not convinced it's actually needed at all. (Also, I'm not entirely sure that a express middleware style function can also be an EventEmitter...)

Reply:
> I figured out what was throwing me off: It sounds like you're wanting to give extra info on the last request that's allowed through - the 3rd request if the limit is 3. Is that right?

Yes.

> However, onLimitReached is triggered on the 4th request

Ohk, I didn't realise that 🤦🏾 

> I think your use-case would be better served by checking the req.rateLimit property after express-rate-limit has done it's thing.

👍🏾 

> A better replacement might be to return an EventEmitter and have some kind of event for this, but I'm not convinced it's actually needed at all. (Also, I'm not entirely sure that a express middleware style function can also be an EventEmitter...)

An `EventEmmitter` would make it reallyy complicated 😅 , and as you mentioned, this kind of use case is already covered in the `handler` function. 

> I think we should keep it in there with the same behavior for now, but just mark it as deprecated try to discourage people from using it

So I'll add a `@deprecated` tag to the `onLimitReached` method but the middleware will still continue to support it.